test/e2e/operator/validate/validate.go

```diff
@@ -21,75 +21,181 @@ import (
+			endCount, err := ready()
+			if err == nil {
+				return endCount, nil
+			}
+			lastErr = err
+		case <-ctx.Done():
+			return 0, fmt.Errorf("timed out waiting for deployment metric, last error: %v", lastErr)
+		}
+	}
+}
 
-	if deployment.GetUID() == "" {
-		return errors.New("deployment not an object that was stored in the API server, no UID")
+// CheckForObjectUpdates wait and checks for deployed driver components
+// in consistent state. That means no unnecessary updates occurred and all
+// expected object updates passed in expectedUpdates have occurred.
+// It uses the 'pmem_csi_deployment_sub_resource_updated_at' metric.
+//
+// When the CR just got created, the operator should
+// immediately create objects with the right content and then
+// not update them again unless it is an expected update from the
+// caller.
+func CheckForObjectUpdates(ctx context.Context, c *deploy.Cluster, metricsURL string, expectedUpdates []client.Object, deployment *api.PmemCSIDeployment) error {
+	type updateInfo struct {
+		expectedObjectLabels []map[string]string
+		isFound              []bool
 	}
 
-	// Track resource versions to detect modifications?
-	var resourceVersions map[string]string
-	if initialCreation {
-		resourceVersions = map[string]string{}
+	info := updateInfo{}
+	for _, o := range expectedUpdates {
+		info.expectedObjectLabels = append(info.expectedObjectLabels, metrics.GetSubResourceLabels(o))
+		info.isFound = append(info.isFound, false)
 	}
 
-	// TODO: once there is a better way to detect that the
-	// operator has finished updating the deployment, check for
-	// that here instead of repeatedly checking the objects.
-	// As it stands now, permanent differences will only be
-	// reported when the test times out.
-	ready := func() (final bool, err error) {
-		return DriverDeployment(ctx, client, k8sver, namespace, deployment, resourceVersions)
+	// isUpdateExpected check if the given object labels match with the
+	// expectedObjectLabels,
+	isUpdateExpected := func(labels map[string]string) bool {
+		for i, ol := range info.expectedObjectLabels {
+			if reflect.DeepEqual(ol, labels) {
+				info.isFound[i] = true
+				return true
+			}
+		}
+		return false
 	}
-	if final, err := ready(); err != nil {
-		if final {
+
+	checkObjectUpdates := func() error {
+		mf, err := deploy.GetMetrics(ctx, c, metricsURL)
+		if err != nil {
 			return err
 		}
-	loop:
-		for {
-			select {
-			case <-ticker.C:
-				final, err = ready()
-				if err == nil {
-					break loop
-				}
-				if final {
-					return err
-				}
-			case <-ctx.Done():
-				return fmt.Errorf("timed out waiting for deployment, last error: %v", err)
+		updates, ok := mf["pmem_csi_deployment_sub_resource_updated_at"]
+		if !ok {
+			return nil
+		}
+
+		unExpectedList := []string{}
+		for _, m := range updates.GetMetric() {
+			lblMap := labelPairToMap(m.GetLabel())
+			if strings.Contains(lblMap["ownedBy"], string(deployment.UID)) && !isUpdateExpected(lblMap) {
```

Comment:
Nit: why is this a substring search and not a direct comparison?

I'll merge the PR despite this, but it looks odd.